src/map.rs

```diff
@@ -1348,6 +1348,38 @@ where
         self.put(key, value, false, guard)
     }
 
+    #[inline]
+    /// Tries to insert a key-value pair into the map.
+    ///
+    /// If the map did not have this key present, [`None`] is returned and the
+    /// key-value pair is inserted into the map.
+    ///
+    /// If the map did have this key present, and the old value is returned.
+    /// Neither the key or value are updated; this matters for types that can
+    /// be `==` without being identical. See the [std-collections
+    /// documentation] for more.
+    ///
+    /// [`None`]: std::option::Option::None
+    /// [std-collections documentation]: https://doc.rust-lang.org/std/collections/index.html#insert-and-complex-keys
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use flurry::HashMap;
+    ///
+    /// let map = HashMap::new();
+    /// let mref = map.pin();
+    ///
+    /// mref.insert(37, "a");
+    /// assert_eq!(mref.try_insert(37, "b"), Some(&"a"));
+    /// assert_eq!(mref.try_insert(42, "c"), None);
+    /// assert_eq!(mref.get(&37), Some(&"a"));
+    /// assert_eq!(mref.get(&42), Some(&"c"));
+    /// ```
+    pub fn try_insert<'g>(&'g self, key: K, value: V, guard: &'g Guard) -> Option<&'g V> {
```

Comment:
Regarding the return type and https://github.com/jonhoo/flurry/issues/58#issuecomment-602286897, I think `Result<(), &'g V>` is fine. The existing `try_XYZ` implementations differ in what they return as `Ok(_)`, depending on you would expect a return value from the "regular" version of these operations. For example, `Mutex::try_lock` returns `Ok(Guard)`, where `Guard` can then be used to access the object behind the Mutex. But `HashMap::try_reserve` returns `Ok(())`, since you'd probably not expect a return when reserving some space.

In our case, the regular version of `try_insert` is simply `insert`, where the key is not present in the map yet. The return of `insert` in that case is `None`, which is about as informative as `()` if you know the key can't have been there before ^^

Reply:
Seems reasonable.

Reply:
See also my argument in https://github.com/jonhoo/flurry/issues/58#issuecomment-602568241 for having `Err` be:
```rust
struct TryInsertError {
    current: &T,
    new: T,
}
```

Reply:
This would still necessitate changes to `put`, right? Because ownership of the new value is transferred to it (so it can create a new node) and we would need to get it back.

Reply:
> This would still necessitate changes to `put`, right? Because ownership of the new value is transferred to it (so it can create a new node) and we would need to get it back.

Exactly. That's the only way for us to return any information about the inserted value in the success case is to change `put`'s type signature.

Reply:
I personally feel like `Result<(), &V>` is fine.

Reply:
Yup, it would, which I think is reasonable. We'd change `put` to have an (internal) `enum` return type that indicates the outcome of the `put` call.

Reply:
I could whip up an updated `put` quickly. Give me a sec.

Reply:
See https://github.com/jonhoo/flurry/pull/76